File: cpp/src/max_subarray_length_2958.cpp
// https://leetcode.com/problems/length-of-longest-subarray-with-at-most-k-frequency

#include "test.hpp"

#include <algorithm>
#include <unordered_map>
#include <vector>

    int maxSubarrayLength(const std::vector<int>& nums, int k) {
        std::unordered_map<int, int> freq;
        // Index of the next element leaving the window.
        std::size_t back = 0;
        int length = 0;
        int longest_length = 0;
        for (int num : nums) {
            int& bucket = freq[num];
            ++bucket;
            ++length;
            if (bucket > k) {
                // Only `num` can exceed `k`, and by exactly one, so shrinking
                // until its earliest occurrence leaves the window is enough.
                while (true) {
                    int evicted = nums[back];
                    ++back;
                    --freq[evicted];
                    --length;
                    if (evicted == num) {
                        break;
                    }
                }
            }
            longest_length = std::max(longest_length, length);
        }
        return longest_length;
    }

TEST(official1) {
    check_eq(6, maxSubarrayLength({1, 2, 3, 1, 2, 3, 1, 2}, 2));
}

TEST(official2) {
    check_eq(2, maxSubarrayLength({1, 2, 1, 2, 1, 2, 1, 2}, 1));
}

TEST(official3) {
    check_eq(4, maxSubarrayLength({5, 5, 5, 5, 5, 5, 5}, 4));
}

// The officials all evict at least once, so they never measure a window
// that still starts at index 0. These do.
TEST(single_element) {
    check_eq(1, maxSubarrayLength({1}, 1));
}

TEST(whole_array_is_good) {
    check_eq(3, maxSubarrayLength({1, 2, 3}, 3));
}

TEST(longest_window_is_a_prefix) {
    check_eq(4, maxSubarrayLength({2, 3, 1, 1, 1}, 2));
}

TEST(longest_window_is_a_suffix) {
    check_eq(4, maxSubarrayLength({1, 1, 1, 2, 3, 1, 1}, 2));
}

File: rust/src/max_subarray_length_2958.rs
// https://leetcode.com/problems/length-of-longest-subarray-with-at-most-k-frequency

use std::collections::HashMap;

pub fn max_subarray_length(nums: Vec<i32>, k: i32) -> i32 {
    assert!(k >= 1);
    let k: usize = k.try_into().unwrap();
    let mut freq = HashMap::new();
    let mut front = nums.iter().enumerate();
    let mut back = nums.iter().enumerate();
    let mut longest_length = 0;
    let mut back_position = 0;
    let mut front_position ;
    'outer: loop {
        loop {
            let Some((i, &num)) = front.next() else {
                break 'outer;
            };
            let bucket: &mut usize = freq.entry(num).or_default();
            *bucket += 1;
            front_position = i;
            if *bucket > k {
                break;
            }
            longest_length = longest_length.max(front_position - back_position);
        }
        debug_assert_eq!(1, freq.values().filter(|&&v| v > k).count());
        loop {
            let Some((i, &num)) = back.next() else {
                break 'outer;
            };
            back_position = i;
            let bucket = freq.get_mut(&num).unwrap();
            *bucket -= 1;
            if *bucket == k {
                longest_length = longest_length.max(front_position - back_position);
                break;
            }
        }
        debug_assert_eq!(0, freq.values().filter(|&&v| v > k).count());
    }
    i32::try_from(longest_length).expect("longest_length should be less than i32::MAX but was: {longest_length}")
}

#[cfg(test)]
mod tests {
    use super::max_subarray_length;

    #[test]
    fn official1() {
        assert_eq!(6, max_subarray_length(vec![1, 2, 3, 1, 2, 3, 1, 2], 2));
    }

    #[test]
    fn official2() {
        assert_eq!(2, max_subarray_length(vec![1, 2, 1, 2, 1, 2, 1, 2], 1));
    }

    #[test]
    fn official3() {
        assert_eq!(4, max_subarray_length(vec![5, 5, 5, 5, 5, 5, 5], 4));
    }
}
